jax/flatten_util.py

```diff
@@ -26,18 +31,42 @@ def ravel_pytree(pytree):
   """Ravel (i.e. flatten) a pytree of arrays down to a 1D array.
 
   Args:
-    pytree: a pytree to ravel.
+    pytree: a pytree of arrays and scalars to ravel.
 
   Returns:
     A pair where the first element is a 1D array representing the flattened and
-    concatenated leaf values, and the second element is a callable for
-    unflattening a 1D vector of the same length back to a pytree of of the same
-    structure as the input ``pytree``.
+    concatenated leaf values, with dtype determined by promoting the dtypes of
+    leaf values, and the second element is a callable for unflattening a 1D
+    vector of the same length back to a pytree of of the same structure as the
+    input ``pytree``.
+
+  For details on dtype promotion, see
+  https://jax.readthedocs.io/en/latest/type_promotion.html.
   """
   leaves, treedef = tree_flatten(pytree)
-  flat, unravel_list = vjp(_ravel_list, *leaves)
+  dts = [dtypes.dtype(l) for l in leaves]
+  dtype = functools.reduce(dtypes.promote_types, dts)
```

Comment:
It would be more efficient here to call `dtypes.promote_types(*dts)`, and the lattice property ensures that it will give the same results 😎

Reply:
That would be nice, but iiuc [`dtypes.promote_types` only accepts two arguments](https://github.com/google/jax/blob/d75becbf676e392bcf08e91594e8495bbc39317d/jax/dtypes.py#L299), though `_least_upper_bound` accepts many.

Maybe I should call `_least_upper_bound` directly? What do you advise?

Reply:
Oh, sorry, I meant to say `dtypes.result_type(*dts)`

Reply:
I'm not sure why `promote_types` is a separate implementation...

Reply:
Okay, I'll try that!